### PR TITLE
[Ameba] Update BLEManagerImpl for function extension

### PR DESCRIPTION
**Pull request by pankore**
- Move platform related operations to platform SDK.
- Undefine CHIP_DEVICE_CONFIG_CHIPOBLE_ENABLE_ADVERTISING_AUTOSTART.

### CODE REVIEW

File: src/platform/Ameba/BLEManagerImpl.cpp [hunk 1]
```diff
@@ -810,6 +851,169 @@ bool BLEManagerImpl::IsSubscribed(uint16_t conId)
...
+    // Post an event to the Chip queue to deliver the data into the Chip stack.
+    {
+        ChipDeviceEvent event;
+        event.Type                        = DeviceEventType::kCHIPoBLEWriteReceived;
+        event.CHIPoBLEWriteReceived.ConId = (uint16_t) conn_id;
+        event.CHIPoBLEWriteReceived.Data  = std::move(buf).UnsafeRelease();
+        PlatformMgr().PostEventOrDie(&event);
+    }
+
+    if (err != CHIP_NO_ERROR)
```
Comment (andy31415) [inline src/platform/Ameba/BLEManagerImpl.cpp:870]: remove err usage: it is never set to anything except CHIP_NO_ERROR

File: src/platform/Ameba/BLEManagerImpl.cpp [hunk 2]
```diff
@@ -810,6 +851,169 @@ bool BLEManagerImpl::IsSubscribed(uint16_t conId)
...
+        ChipLogError(DeviceLayer, "HandleRXCharWrite() failed: %s", ErrorStr(err));
+    }
+}
+
+#if defined(CONFIG_MATTER_BLEMGR_ADAPTER) && CONFIG_MATTER_BLEMGR_ADAPTER
+CHIP_ERROR BLEManagerImpl::matter_blemgr_gap_connect_cb(uint8_t conn_id)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    err            = sInstance.HandleGAPConnect((uint16_t) conn_id);
+    SuccessOrExit(err);
```
Comment (andy31415) [inline src/platform/Ameba/BLEManagerImpl.cpp:881]: there seems to be no need for a `goto exit` pattern here.

File: src/platform/Ameba/BLEManagerImpl.cpp [hunk 3]
```diff
@@ -810,6 +851,169 @@ bool BLEManagerImpl::IsSubscribed(uint16_t conId)
...
+    PlatformMgr().ScheduleWork(DriveBLEState, 0);
+
+    return err;
+}
+
+CHIP_ERROR BLEManagerImpl::matter_blemgr_gap_disconnect_cb(uint8_t conn_id, uint16_t disc_cause)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    err            = sInstance.HandleGAPDisconnect((uint16_t) conn_id, disc_cause);
+    SuccessOrExit(err);
```
Comment (andy31415) [inline src/platform/Ameba/BLEManagerImpl.cpp:900]: remove the `SuccessOrExit` and `exit:` label here as well: there seems no need for it.

File: src/platform/Ameba/BLEManagerImpl.cpp [hunk 4]
```diff
@@ -810,6 +851,169 @@ bool BLEManagerImpl::IsSubscribed(uint16_t conId)
...
+void BLEManagerImpl::matter_blemgr_rx_char_write_cb(uint8_t conn_id, uint8_t * p_value, uint16_t len)
+{
+    sInstance.HandleRXCharWrite(p_value, len, conn_id);
+
+    PlatformMgr().ScheduleWork(DriveBLEState, 0);
+}
+
+void BLEManagerImpl::matter_blemgr_tx_char_cccd_write_cb(uint8_t conn_id, uint8_t indicationsEnabled, uint8_t notificationsEnabled)
+{
+    sInstance.HandleTXCharCCCDWrite((int) conn_id, (int) indicationsEnabled, (int) notificationsEnabled);
```
Comment (andy31415) [inline src/platform/Ameba/BLEManagerImpl.cpp:924]: prefer C++ casts to C style casts

File: src/platform/Ameba/BLEManagerImpl.cpp [hunk 5]
```diff
@@ -755,6 +788,14 @@ void BLEManagerImpl::DriveBLEState(intptr_t arg)
     sInstance.DriveBLEState();
 }
 
+void BLEManagerImpl::InitSubscribed(void)
```
Comment (andy31415) [inline src/platform/Ameba/BLEManagerImpl.cpp:791]: remove `(void)` argument and just use `()`

File: src/platform/Ameba/BLEManagerImpl.cpp [hunk 6]
```diff
@@ -810,6 +851,169 @@ bool BLEManagerImpl::IsSubscribed(uint16_t conId)
...
+    }
+    break;
+    default:
+        break;
+    }
+
+    return 0;
+}
+#else
+#if CHIP_ENABLE_ADDITIONAL_DATA_ADVERTISING
```
Comment (andy31415) [inline src/platform/Ameba/BLEManagerImpl.cpp:982]: could we use `#elif` ? given that the block of code is large, could we comment to what else branch this corresponds to (like `// not defined(CONFIG_MATTER_BLEMGR_ADAPTER) && CONFIG_MATTER_BLEMGR_ADAPTER`)?

Why is the additional data advertising depending on BLEMGR adapter not being used? would that option not apply to any type of BLE support?

File: src/platform/Ameba/BLEManagerImpl.cpp [hunk 7]
```diff
@@ -1019,7 +1166,7 @@ int BLEManagerImpl::ble_callback_dispatcher(void * param, void * p_cb_data, int
     }
     return 0;
 }
-
+#endif
```
Comment (andy31415) [inline src/platform/Ameba/BLEManagerImpl.cpp:1169]: add a comment what if this ends.

File: src/platform/Ameba/BLEManagerImpl.h [hunk 8]
```diff
@@ -156,15 +148,27 @@ class BLEManagerImpl final : public BLEManager, private BleLayer, private BlePla
...
     BLEManagerImpl::CHIPoBLEConState * GetConnectionState(uint8_t connectionHandle, bool allocate);
+#if defined(CONFIG_MATTER_BLEMGR_ADAPTER) && CONFIG_MATTER_BLEMGR_ADAPTER
+    static CHIP_ERROR matter_blemgr_gap_connect_cb(uint8_t conn_id);
+    static CHIP_ERROR matter_blemgr_gap_disconnect_cb(uint8_t conn_id, uint16_t disc_cause);
+    static void matter_blemgr_rx_char_write_cb(uint8_t conn_id, uint8_t * p_value, uint16_t len);
+    static void matter_blemgr_tx_char_cccd_write_cb(uint8_t conn_id, uint8_t indicationsEnabled, uint8_t notificationsEnabled);
+    static CHIP_ERROR matter_blemgr_tx_complete_cb(uint8_t conn_id);
+    static int matter_blemgr_callback_dispatcher(void * param, T_MATTER_BLEMGR_CALLBACK_TYPE cb_type, void * p_cb_data);
+#else
+#if CHIP_ENABLE_ADDITIONAL_DATA_ADVERTISING
```
Comment (andy31415) [inline src/platform/Ameba/BLEManagerImpl.h:164]: add a comment why additional data advertisement does not work or is not needed with matter blemgr adapter

File: src/platform/Ameba/BLEManagerImpl.cpp (hunk 4 above)
Comment (pankore) [reply src/platform/Ameba/BLEManagerImpl.cpp]: Done

File: src/platform/Ameba/BLEManagerImpl.cpp (hunk 6 above)
Comment (pankore) [reply src/platform/Ameba/BLEManagerImpl.cpp:1023]: Fixed

File: src/platform/Ameba/BLEManagerImpl.cpp (hunk 7 above)
Comment (pankore) [reply src/platform/Ameba/BLEManagerImpl.cpp]: Comment added

File: src/platform/Ameba/BLEManagerImpl.h (hunk 8 above)
Comment (pankore) [reply src/platform/Ameba/BLEManagerImpl.h:168]: Fixed, it should also work with matter_blemgr_adapter.

**Comment by pankore** [comment]
Hi [andy31415](https://github.com/andy31415), [anush-apple](https://github.com/anush-apple), [arkq](https://github.com/arkq), [Byungjoo-Lee](https://github.com/Byungjoo-Lee), [bzbarsky-apple](https://github.com/bzbarsky-apple), [carol-apple](https://github.com/carol-apple), [chrisdecenzo](https://github.com/chrisdecenzo), [chshu](https://github.com/chshu), [chulspro](https://github.com/chulspro), [Damian-Nordic](https://github.com/Damian-Nordic), [dhrishi](https://github.com/dhrishi), [electrocucaracha](https://github.com/electrocucaracha), [emargolis](https://github.com/emargolis), [franck-apple](https://github.com/franck-apple), [gjc13](https://github.com/gjc13), [harimau-qirex](https://github.com/harimau-qirex), [harsha-rajendran](https://github.com/harsha-rajendran), [hawk248](https://github.com/hawk248), [jelderton](https://github.com/jelderton), [jepenven-silabs](https://github.com/jepenven-silabs), [jmartinez-silabs](https://github.com/jmartinez-silabs), [jtung-apple](https://github.com/jtung-apple), [ksperling-apple](https://github.com/ksperling-apple), [lazarkov](https://github.com/lazarkov), [lpbeliveau-silabs](https://github.com/lpbeliveau-silabs), [LuDuda](https://github.com/LuDuda), [mlepage-google](https://github.com/mlepage-google), [msandstedt](https://github.com/msandstedt), [mspang](https://github.com/mspang), [robszewczyk](https://github.com/robszewczyk), [saurabhst](https://github.com/saurabhst), [selissia](https://github.com/selissia), [tcarmelveilleux](https://github.com/tcarmelveilleux), [tecimovic](https://github.com/tecimovic), [turon](https://github.com/turon), [vijs](https://github.com/vijs), [vivien-apple](https://github.com/vivien-apple), [woody-apple](https://github.com/woody-apple), [xylophone21](https://github.com/xylophone21), [yufengwangca](https://github.com/yufengwangca), [yunhanw-google](https://github.com/yunhanw-google),

Please kindly help to review our PR, many thanks!

**Review by andy31415 — APPROVED** [review]

**Pull request merged**